File: leetcode_75/Queue/python/649.py
from collections import deque


class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        n = len(senate)
        queue_R = deque()
        queue_D = deque()
        for i in range(n):
            if senate[i] == "R":
                queue_R.append(i)
            else:
                queue_D.append(i)

        while queue_R and queue_D:
            r_index = queue_R.popleft()
            d_index = queue_D.popleft()
            if r_index < d_index:
                queue_R.append(r_index + n)
            else:
                queue_D.append(d_index + n)

        return "Radiant" if queue_R else "Dire"


solution = Solution()
print(solution.predictPartyVictory(senate="RD"))

File: leetcode_75/Queue/cpp/649.cpp
#include <string>
#include <queue>

using namespace std;

class Solution
{
public:
    string predictPartyVictory(string senate)
    {
        int n = senate.size();
        queue<int> queue_R;
        queue<int> queue_D;

        for (int i = 0; i < n; ++i)
        {
            if (senate[i] == 'R')
            {
                queue_R.push(i);
            }
            else
            {
                queue_D.push(i);
            }
        }

        while (queue_R.size() > 0 && queue_D.size() > 0)
        {
            int r_index = queue_R.front();
            queue_R.pop();
            int d_index = queue_D.front();
            queue_D.pop();

            if (r_index < d_index)
            {
                queue_R.push(r_index + n);
            }
            else
            {
                queue_D.push(d_index + n);
            }
        }

        return queue_R.size() > 0 ? "Radiant" : "Dire";
    }
};